app/screens/entry/entry.js

```diff
@@ -0,0 +1,272 @@
+import {stripTrailingSlashes} from 'app/utils/url';
+import {makeStyleSheetFromTheme} from 'app/utils/theme';
+
+
+import EmptyToolbar from 'app/components/start/empty_toolbar';
+import Loading from 'app/components/loading';
+
+const lazyLoadSelectServer = () => {
+    return require('app/screens/select_server').default;
+};
+
+const lazyLoadChannel = () => {
+    return require('app/screens/channel').default;
+};
+
+const lazyLoadPushNotifications = () => {
+    return require('app/utils/push_notifications').configurePushNotifications;
+};
+
+const lazyLoadReplyPushNotifications = () => {
+    return require('app/utils/push_notifications').onPushNotificationReply;
+};
+
+/**
+ * Entry Component:
+ * With very little overhead navigate to
+ *  - Login or
+ *  - Channel Component
+ *
+ * The idea is to render something to the user as soon as possible
+ */
+export default class Entry extends PureComponent {
+    static propTypes = {
+        config: PropTypes.object,
+        theme: PropTypes.object,
+        navigator: PropTypes.object,
+        isLandscape: PropTypes.bool,
+        hydrationComplete: PropTypes.bool,
+        initializeModules: PropTypes.func.isRequired,
+        actions: PropTypes.shape({
+            setDeviceToken: PropTypes.func.isRequired,
+        }).isRequired,
+    };
+
+    constructor(props) {
+        super(props);
+
+        this.state = {
+            launchLogin: false,
+            launchChannel: false,
+        }
+    }
+
+    componentDidMount() {
+        Client4.setUserAgent(DeviceInfo.getUserAgent());
+        this.unsubscribeFromStore = store.subscribe(this.listenForHydration);
+
+        EventEmitter.on(ViewTypes.LAUNCH_LOGIN, this.handleLaunchLogin);
+        EventEmitter.on(ViewTypes.LAUNCH_CHANNEL, this.handleLaunchChannel);
+    }
+
+    componentWillUnmount() {
+        EventEmitter.off(ViewTypes.LAUNCH_LOGIN, this.handleLaunchLogin);
+        EventEmitter.off(ViewTypes.LAUNCH_CHANNEL, this.handleLaunchChannel);
+    }
+
+    handleLaunchLogin = (initializeModules) => {
+      this.setState({launchLogin: true});
+
+      if (initializeModules) {
+          this.props.initializeModules();
+      }
+    };
+
+    handleLaunchChannel = (initializeModules) => {
+        this.setState({launchChannel: true});
+
+        if (initializeModules) {
+            this.props.initializeModules();
+        }
+    };
+
+    listenForHydration = () => {
+        const {getState} = store;
+        const state = getState();
+
+        if (!app.isNotificationsConfigured) {
+            this.configurePushNotifications();
+        }
+
+        if (state.views.root.hydrationComplete) {
+            this.unsubscribeFromStore();
+
+            this.setAppCredentials();
+            this.setReplyNotifications();
+
+            if (Platform.OS === 'android') {
+                this.launchForAndroid();
+                return;
+            }
+
+            this.launchForiOS();
+        }
+    };
```

Comment:
At this point the only useful thing Entry component does is wait for `rehydration`

That's something that may move to `mattermost.android.js` in order to subscribe to the store sooner.  Before the component even renders, there's an overhead to start a new Activity so that may be a good next change.

For the meantime, it works just fine in this class